### ISSUE TITLE
[display] @signature with function argument

### ISSUE THREAD

**Issue by Gama11** [HaxeFoundation/haxe#6383]
Request the `@signature` at the single `(` here returns the signature of `sort()`, even though it's inside of a function declaration (vshaxe always requests the signature on `(` or `,`). 

```haxe
class Main {
	public static function main() {
		var list:Array<String> = [];
		list.sort(function(a, b) {
			(
			return 0;
		});
	}
}
```

```
haxe --display Main.hx@115@signature
```

![](http://i.imgur.com/OD2bLWD.png)

**Comment by nadako** [comment]
I think this is correct behaviour from Haxe: we request signature and we get it. While it may look funny in this example, in a case where you explicitly request signature info (with ctrl+shift+space by default in VSCode), Haxe must return that even if we're within a callback function that is used as an argument.

So this looks more like a vshaxe issue: it should just not request signature completion on any `(`, but only if it's a function call somehow.

**Comment by Simn** [comment]
I agree. If you request the signature and your cursor is in a function argument, you get the signature for the function that holds said function argument. I don't think we should start introducing special cases here, so this has to be fixed in vshaxe.

**Comment by Gama11** [comment]
Haxe can help here if vshaxe tells it how completion was invoked, but that needs VSCode API support: https://github.com/Microsoft/vscode/issues/34737

**Comment by Simn** [comment]
Blocked on vshaxe reporting the correct `wasAutoTriggered` value.

**Comment by Gama11** [comment]
vshaxe currently already sends `wasAutoTriggered: true`, which means Haxe still isn't behaving as it should.

**Comment by Simn** [comment]
Yes but if I fix it then signature completion is broken for all cases where the caret isn't at the call `(`.

**Comment by Gama11** [comment]
Well.. we could change it to `false` in vshaxe for now.

**Comment by Simn** [comment]
See linked commit. If I understand the problem correctly, we want to have an exact position check if auto-triggered and the current range on otherwise. If you want to try this with a different default you can just remove the `true ||`.

**Comment by Gama11** [comment]
The test case works as expected with that (no signature help on that random `(` if auto-triggered). However, I noticed that signature help now closes whenever you type something other than one of the trigger characters (like a space after a comma), which is not ideal...

Perhaps this is why [the API they've recently added to VSCode is a tri-state](https://github.com/Microsoft/vscode/pull/58135):

```ts
export enum SignatureHelpTriggerReason {
	Invoke = 1,
	TriggerCharacter = 2,
	Retrigger = 3,
}
```

If I understand the semantics of `Retrigger` correctly, the language server could probably just send Haxe `wasAutoTriggered: triggerReason == TriggerCharacter`. The `wasAutoTriggered` naming seems a bit confusing then though.

**Comment by Simn** [comment]
Either way, I guess I should wait with this change till the next vshaxe version is released.

**Comment by Simn** [comment]
The request still has `"wasAutoTriggered":true` even with explicit ctrl shift space.

**Comment by Gama11** [comment]
Yeah, we haven't un-hardcoded it yet (we could, but we need to check how this affects previous Haxe versions we still want to support). https://github.com/vshaxe/haxe-language-server/blob/24f35a2e08b11ade989f733f18112c4c0a9aeb98/src/haxeLanguageServer/features/SignatureHelpFeature.hx#L45

**Comment by Gama11** [comment]
The language server now sends real values. Note that I had to update a VSCode lib, so you need to do an `npm install` for it to work.

**Comment by Gama11** [comment]
I just checked what the situation is like when removing the hardcoded `true ||` from `parser.ml`. It looks like auto-trigger with `(` is fixed with that, but still happens with `,`:

![](https://i.imgur.com/hgZUMoW.gif)

Also, I think there's some off-by-one error when moving the cursor outside of the signature. Rather than closing, the signature help stays open and highlights the first parameter again:

![](https://i.imgur.com/35fHUXj.gif)
